Client/Frontend/Home/PanelDataObservers.swift

```diff
@@ -70,12 +70,14 @@ class ActivityStreamDataObserver: DataObserver {
         }
 
         self.delegate?.willInvalidateDataSources()
-        self.profile.recommendations.repopulate(invalidateTopSites: shouldInvalidateTopSites, invalidateHighlights: shouldInvalidateHighlights).uponQueue(queue) { _ in
-            if shouldInvalidateTopSites {
-                self.profile.prefs.setBool(true, forKey: PrefsKeys.KeyTopSitesCacheIsValid)
+        queue.async {
```

Comment:
I think this should happen inside `SQLiteHistoryRecommendations`, on line 147.

Or, indeed, define a sibling method in `BrowserDB.swift`:

```
    /**
     * Runs an array of SQL commands. Note: These will all run in order in a transaction and will block
     * the caller's thread until they've finished. If any of them fail the operation will abort (no more
     * commands will be run) and the transaction will roll back, returning a DatabaseError.
     */
    func run(_ commands: [(sql: String, args: Args?)]) -> Success {
```

that *doesn't* block the calling thread. It simply needs to return a `Deferred` and fill it from the block inside `self.transaction(synchronous: false, …)`.

Then you can just make `repopulate` call `self.db.runAsync` instead of `self.db.run`.